tools/engine_tools_lib/lib/engine_tools_lib.dart

```diff
@@ -0,0 +1,184 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+/// A minimal library for discovering and probing a local engine repository.
+///
+/// This library is intended to be used by tools that need to interact with a
+/// local engine repository, such as `clang_tidy` or `githooks`. For example,
+/// finding the `compile_commands.json` file for the most recently built output:
+///
+/// ```dart
+/// final Engine engine = Engine.findWithin();
+/// final Output? output = engine.latestOutput();
+/// if (output == null) {
+///   print('No output targets found.');
+/// } else {
+///   final io.File? compileCommandsJson = output.compileCommandsJson;
+///   if (compileCommandsJson == null) {
+///     print('No compile_commands.json file found.');
+///   } else {
+///     print('Found compile_commands.json file at ${compileCommandsJson.path}');
+///   }
+/// }
+/// ```
+library;
+
+import 'dart:io' as io;
+
+import 'package:path/path.dart' as p;
+
+/// Represents the `$ENGINE` directory (i.e. a checked-out Flutter engine).
+///
+/// If you have a path to the `$ENGINE/src` directory, use [Engine.fromSrcPath].
+///
+/// If you have a path to a directory within the `$ENGINE/src` directory, or
+/// want to use the current working directory, use [Engine.findWithin].
+final class Engine {
+  /// Creates an [Engine] from a path such as `/Users/.../flutter/engine/src`.
+  ///
+  /// ```dart
+  /// final Engine engine = Engine.findWithin('/Users/.../engine/src');
+  /// print(engine.srcDir.path); // /Users/.../engine/src
+  /// ```
+  ///
+  /// Throws an [ArgumentError] if the path does not point to a valid engine.
+  factory Engine.fromSrcPath(String srcPath) {
+    // If the path does not end in `/src`, fail.
+    if (p.basename(srcPath) != 'src') {
+      throw ArgumentError(
+        'The path $srcPath does not end in `${p.separator}src`.\n'
+        '\n$_kHintFindWithinPath'
+      );
+    }
+
+    // If the directory does not exist, or is not a directory, fail.
+    final io.Directory srcDir = io.Directory(srcPath);
+    if (!srcDir.existsSync()) {
+      throw ArgumentError(
+        'The path "$srcPath" does not exist or is not a directory.\n'
+        '\n$_kHintFindWithinPath'
+      );
+    }
+
+    // Check for the existence of a `flutter` directory within `src`.
+    final io.Directory flutterDir = io.Directory(p.join(srcPath, 'flutter'));
+    if (!flutterDir.existsSync()) {
+      throw ArgumentError(
+        'The path "$srcPath" does not contain a "flutter" directory.'
+      );
+    }
+
+    // Check for the existence of a `out` directory within `src`.
+    final io.Directory outDir = io.Directory(p.join(srcPath, 'out'));
```

Comment:
Could this be optional? `outputs()` below could return an empty list, and `latestOutput()` could return `null`.

Reply:
Yeah, very fair. Updated.